### ISSUE TITLE
Gitlint hangs on Cygwin

### ISSUE THREAD

**Issue by jorisroovers** [jorisroovers/gitlint#91]
This also affects 'git bash' which is just Cygwin with a different name.

**TLDR**: Only way to use gitlint with cygwin (for now) is via stdin:
```sh
git log -1 --pretty="%B" | gitlint
```

When using gitlint on Cygwin, it will hang when trying to determine whether there is data being passed via stdin.

```sh
# This hangs
gitlint
# Note that Other commands do work, like:
gitlint --version
gitlint generate-config
gitlint install-hook
```

The issue here is with how Cygwin emulates a shell in windows. In particular, in the testing I've done, `stdin` is always a named pipe during gitlint invocation.

```sh
# Returns True in Cygwin, False everywhere else
python -c "import sys; import os; import stat; print stat.S_ISFIFO(os.fstat(sys.stdin.fileno()).st_mode)"
```

The issue with this is that once gitlint thinks that stdin is a pipe, it will then attempt and fail to read the stdin(), indefinitely waiting (i.e. gitlint hangs).

The relevant code is here: https://github.com/jorisroovers/gitlint/blob/e95fe6122fae5930534f46dfc15dcd3bb7cf8059/gitlint/cli.py#L100-L129

The only way to work around this right now is actually pass data to stdin, like so:

```sh
git log -1 --pretty="%B" | gitlint
```

I did a bit of research into this and identified a few potential workarounds that need more investigation:

1. Detect CYGWIN environment and print a warning message. The only way to reliably detect cygwin seems to try to read the output of the `uname` command. Python's `platform` module will *NOT* consistently report correctly on Cygwin (if python was installed system-wide outside of Cygwin, [`platform` will report `Windows` and not `Cygwin`](https://stackoverflow.com/a/1387226/381010)).
2. Attempt implementing non-blocking reads on windows: https://stackoverflow.com/questions/34504970/non-blocking-read-on-os-pipe-on-windows
3. Still find a different way to detect that data is not being piped into gitlint, or somehow fix what `stat.S_ISFIFO` returns by finding/making an environment change in Cygwin.

Some additional pointers that might be relevant:
- http://cygwin.1069669.n5.nabble.com/Why-is-stdin-always-a-pipe-td141695.html
- https://stackoverflow.com/questions/41144737/python-input-does-not-detect-eol-in-mingw-terminal-but-does-in-cmd-terminal

**Comment by jorisroovers** [comment]
Additional workaround was added with the `--ignore-stdin` flag in the upcoming v0.12.0 release. Use it like so:
```sh
gitlint --ignore-stdin
```
or in a `.gitlint` file:
```ini
[general]
ignore-stdin=true
```
